File: tutorial_solutions/stack_tuerk.c
struct node {
    struct node *next;
    int value;
};

/*@

predicate nodes(struct node *node, int count) =
    node == 0 ?
        count == 0
    :
        0 < count &*&
        node->next |-> ?next &*& node->value |-> ?value &*& malloc_block_node(node) &*&
        nodes(next, count - 1);

@*/

struct stack {
    struct node *head;
};

/*@

predicate stack(struct stack *stack, int count) =
    stack->head |-> ?head &*& malloc_block_stack(stack) &*& 0 <= count &*& nodes(head, count);

@*/

int stack_get_count(struct stack *stack)
    //@ requires stack(stack, ?count);
    //@ ensures stack(stack, count) &*& result == count;
{
    //@ open stack(stack, count);
    struct node *n = stack->head;
    int i = 0;
    for (;;)
        //@ requires nodes(n, ?count1);
        //@ ensures nodes(old_n, count1) &*& i == old_i + count1;
    {
        //@ open nodes(n, count1);
        if (n == 0) {
            //@ close nodes(n, count1);
            break;
        }
        n = n->next;
        i++;
        //@ recursive_call();
        //@ close nodes(old_n, count1);
    }
    //@ close stack(stack, count);
    return i;
}

File: rust_tutorials/purely_unsafe/solutions/stack_tuerk.rs
// verifast_options{disable_overflow_check}
//@ use std::alloc::{Layout, alloc_block};

struct Node {
    next: *mut Node,
    value: i32,
}

/*@

pred Nodes(node: *mut Node, count: i32) =
    if node == 0 {
        count == 0
    } else {
        0 < count &*&
        (*node).next |-> ?next &*& (*node).value |-> ?value &*&
        alloc_block_Node(node) &*& Nodes(next, count - 1)
    };

@*/

struct Stack {
    head: *mut Node,
}

/*@

pred Stack(stack: *mut Stack, count: i32) =
    (*stack).head |-> ?head &*& alloc_block_Stack(stack) &*& Nodes(head, count);

@*/

unsafe fn stack_get_count(stack: *mut Stack) -> i32
//@ req Stack(stack, ?count);
//@ ens Stack(stack, count) &*& result == count;
{
    //@ open Stack(stack, count);
    let mut n = (*stack).head;
    let mut i = 0;
    loop {
        /*@
        req Nodes(n, ?count1);
        ens Nodes(old_n, count1) &*& i == old_i + count1;
        @*/
        //@ open Nodes(n, count1);
        //@ if n == 0 { close Nodes(n, count1); }
        if n.is_null() {
            break;
        }
        n = (*n).next;
        i += 1;
        //@ recursive_call();
        //@ close Nodes(old_n, count1);
    }
    //@ close Stack(stack, count);
    i
}
